File: testfilenumpy.go
//go:build pythonnumpy

package main

import (
	"fmt"

	"github.com/fabgeyer/goserpent/numpy"
)

// go:pyexport
func PrintDescr(obj *numpy.Array) {
	fmt.Printf("type=%v dims=%d shape=%v\n", obj.Type(), obj.Dims(), obj.Shape())
}

// go:pyexport
func PrintValues(obj *numpy.Array) {
	for val := range obj.Values() {
		fmt.Printf("%v\n", val)
	}
}

// go:pyexport
func PrintIndexedValues(obj *numpy.Array) {
	for idxs, val := range obj.IndexedValues() {
		fmt.Printf("%v = %v\n", idxs, val)
	}
}

// go:pyexport
func AddIntValue(obj *numpy.Array, v int) {
	switch obj.Type() {
	case numpy.NPY_SHORT:
		for idxs, val := range numpy.IndexedValues[int16](obj) {
			obj.SetAt(val+int16(v), idxs...)
		}

	case numpy.NPY_LONG, numpy.NPY_INT:
		for idxs, val := range numpy.IndexedValues[int](obj) {
			obj.SetAt(val+v, idxs...)
		}

	case numpy.NPY_FLOAT:
		for idxs, val := range numpy.IndexedValues[float32](obj) {
			obj.SetAt(val+float32(v), idxs...)
		}

	case numpy.NPY_DOUBLE:
		for idxs, val := range numpy.IndexedValues[float64](obj) {
			obj.SetAt(val+float64(v), idxs...)
		}

	default:
		panic("invalid dtype")
	}
}

File: testfilenumpy.py
import numpy as np
import testmodulenumpy as tmn

x = np.arange(12).reshape(3, 4)
tmn.PrintDescr(x)
tmn.PrintValues(x)

for dtype in [np.float32, np.float64]:
    y = np.random.rand(10, 10).astype(dtype)
    expected = np.copy(y) + 84
    tmn.AddIntValue(y, 84)
    assert np.all(y == expected)

for dtype in [np.int16, np.int32, np.int64]:
    z = np.arange(42).astype(dtype)
    expected = np.copy(z) + 42
    tmn.AddIntValue(z, 42)
    assert np.all(z == expected)
